objects.js

```diff
@@ -8520,6 +8523,48 @@ WatcherMorph.prototype.userMenu = function () {
     return menu;
 };
 
+WatcherMorph.prototype.valueExporter = function (format) {
+    var myself = this,
+        value = this.currentValue,
+        contents,
+        format = format || 'plain',
+        ide = myself.parentThatIsA(IDE_Morph);
+
+    return function () {
+        switch (format) {
+            case 'plain':
+                contents = value instanceof List ? 
+                    value.asArray().join(', ') : 
```

Comment:
Excuse me for barging in with opinions, but, have we discussed 3D lists? :)

---

Aside from which, I *think* I agree with @jguille2 ? I'd go for something like this:

```
NAME,AGE,PLACE
Bernat,32,Barcelona
Joan,41,Barcelona
```

..Simply because it's a list of entries. I can easily iterate through this with a "for each (person) in (all but first of data)".

The above dataset would be represented as this Snap<i>!</i> list, I believe?

```
[
  [NAME, AGE, PLACE],
  [Bernat, 32, Barcelona],
  [Joan, 41, Barcelona]
]
```

I guess you *could* say that a 1D format could be a list of data represented as something like this:

```
TEMPERATURE,27,25,26,22,24,26,25,30, (etc)
```

(First item optional, of course.) But here's the thing – that would *have* to be imported as this:

```
[
  [TEMPERATURE, 27, 25, 26, ...]
]
```

You see, there's a list around the real list containing our data.

So, why not just make it import as `[TEMPERATURE, 27, 25, 26]`, with the useful list on top? The simple answer can be gotten from our table of people – suppose it has just one entry:

```
Brian,67,America
```

If we were to import all 1D CSV data as a single list, it would look like this: `[Brian, 67, America]`; but as soon as we added another item, we'd get this: `[[Brian, 67, America], [Bernat, 32, Barcelona]]`.  And so we suddenly jumped from a 1D list of items to a 2D list of person-entries.

So, at that point, it would probably be easiest to simply break data along newlines, like so:

```
TEMPERATURE
24
28
27
26
25
(etc)
```

---

But I wonder, what should happen if each (or even just some) of the rows contain only one item?

Usually, especially in a CSV table, every row will follow the same format, such as this one:

```
NAME
Joe
Jake
John
Jens
```

But that's not *always* true – take this, for example:

```
NAME,FRUITS..
Joe,Apple,Banana
Jake,Dragonfruit,Pineapple,Passionfruit
John
Jens,Fractal,Orange
```

..John had no fruits. You can see the problem more clearly if you don't have the header row:

```
Joe,Apple,Banana
Jake,Dragonfruit,Pineapple,Passionfruit
John
Jens,Fractal,Orange
```

So the question is – should John be imported as `[John]`, or `John`?

Perhaps the answer is `[John]`, because some other people have fruits. But if *nobody* else has fruits, then you get the imported array of `[Joe, Jake, John, Jens]`, instead of `[[Joe], [Jake], [John], [Jens]]`. A header row could be handy, in this case, then, since it would count as a row that contains more than one item, hence forcing the other rows to be seen as lists containing only one item.

---

Half the problem is, of course, just maintaining the data structure. If I give you this:

```
LIST OF (CSV OF ([
  [1, 2, 3],
  [a, b, c],
  [x, y, z]
]))
```

..I expect to get `[[1, 2, 3], [a, b, c], [x, y, z]]`. If I give you this:

```
LIST OF (CSV OF ([
  [Joe, Apple, Banana],
  [Jake, Dragonfruit, Pineapple, Passionfruit],
  [John],
  [Jens, Fractal, Orange]
]))
```

..I expect to get `[[Joe, Apple, Banana], [Jake, Dragonfruit, Pineapple, Passionfruit], [John], [Jens, Fractal, Orange]]`.

In my humble opinion, JSON is *totally* the way to go, since it just makes everything make more sense – what if I give you `CSV OF [[Joe, [Apple, Banana]], [Jake, [Dragonfruit, ...]], ...]`? 3D lists are tough, but not with JSON, as you likely know! I do know that JSON is, though portable for programming languages, not so useful for human beings, though. Or regex'ers.

---

Double sorry for such a long post, especially when most of this has probably already been discussed; hopefully it at least makes my opinions and input accessible, all at once.

This isn't really an argument against anyone's ideas, just a collected stream of my input!

TL;DR:

* Comma (or semicolon 😛 ) column delimiters and newline row delimiters work well, since they'll provide you with a table of data that's simple to iterate, and easy to read the CSV of, as a human.
* Make sure data structure is maintained; CSV can do 2D arrays so it's reasonable to think giving the CSV block a 2D list or the LIST block the result of that will work.
* 1-or-greater (and 0-or-greater!) type entries can be messy, keep an eye out for those.
* N-dimensional lists are hard, unless you use JSON, but JSON's not really human-friendly.